### PR TITLE
[API9] World Generation

### PR DESCRIPTION
**Pull request by Faithcaio**
**SpongeAPI** | [Sponge](https://github.com/SpongePowered/Sponge/pull/3572)

API I did not add but to consider for the future:
 - Building custom `SurfaceRule`s and exposing all of the vanilla provided SurfaceRules.
   -  I only exposed the SurfaceRules for the main dimensions for now.
 - Building custom terrain `Shaper`s based on CubicSplines like vanilla.

### CODE REVIEW

File: src/main/java/org/spongepowered/api/world/generation/config/SurfaceRules.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package org.spongepowered.api.world.generation.config;
+
+import org.spongepowered.api.Sponge;
+
+public interface SurfaceRules {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/world/generation/config/SurfaceRules.java]: Just `SurfaceRule`

File: src/main/java/org/spongepowered/api/world/generation/config/noise/Shaper.java [hunk 2]
```diff
@@ -22,35 +22,52 @@
...
-import java.util.Objects;
+public interface Shaper {
 
-public interface MultiNoiseConfig {
+    static Shaper overworld() {
+        return Sponge.game().factoryProvider().provide(Shaper.Factory.class).overworld(false);
+    }
+
+    static Shaper amplifiedOverworld() {
+        return Sponge.game().factoryProvider().provide(Shaper.Factory.class).overworld(true);
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/world/generation/config/noise/Shaper.java:36]: Why does toggling overworld true mean amplified? If so, rename this, this is very confusing.

File: src/main/java/org/spongepowered/api/world/generation/config/NoiseGeneratorConfig.java [hunk 3]
```diff
@@ -86,29 +94,31 @@ static Builder builder() {
 
         Builder defaultFluid(BlockState fluid);
 
+        Builder surfaceRule(SurfaceRules rule);
+
         Builder seaLevel(int y);
 
+        Builder mobGeneration(boolean disableMobGeneration);
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/world/generation/config/NoiseGeneratorConfig.java]: Inversed boolean?

File: src/main/java/org/spongepowered/api/world/generation/config/NoiseGeneratorConfig.java [hunk 4]
```diff
@@ -76,6 +82,8 @@ static Builder builder() {
 
     boolean noodleCaves();
 
+    boolean legacyRandomSource();
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/world/generation/config/NoiseGeneratorConfig.java:174]: Is there any point to expose this?

File: src/main/java/org/spongepowered/api/world/generation/config/NoiseGeneratorConfig.java [hunk 5]
```diff
@@ -86,29 +94,31 @@ static Builder builder() {
...
         Builder aquifers(boolean enableAquifers);
 
         Builder noiseCaves(boolean enableNoiseCaves);
 
         Builder oreVeins(boolean enableOreVeins);
 
-        Builder noodleCaves(boolean noodleCaves);
+        Builder noodleCaves(boolean enableNoodleCaves);
+
+        Builder randomSource(boolean useLegacyRandomSource);
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/world/generation/config/NoiseGeneratorConfig.java:207]: See question above regarding the getter.

File: src/main/java/org/spongepowered/api/world/generation/config/SurfaceRules.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
+package org.spongepowered.api.world.generation.config;
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/world/generation/config/SurfaceRules.java]: header

File: src/main/java/org/spongepowered/api/world/generation/config/NoiseGeneratorConfig.java (hunk 3 above)
Comment (Faithcaio) [reply src/main/java/org/spongepowered/api/world/generation/config/NoiseGeneratorConfig.java]: it mirrors what vanilla has in its datapacks

File: src/main/java/org/spongepowered/api/world/generation/config/NoiseGeneratorConfig.java (hunk 4 above)
Comment (Faithcaio) [reply src/main/java/org/spongepowered/api/world/generation/config/NoiseGeneratorConfig.java:174]: From what I can see e.g. nether and end are using this but overworld does not.

File: src/main/java/org/spongepowered/api/world/generation/config/NoiseGeneratorConfig.java (hunk 3 above)
Comment (Zidane) [reply src/main/java/org/spongepowered/api/world/generation/config/NoiseGeneratorConfig.java]: Perhaps but is still so confusing...I'd rather have the API be straightforward then graduate up this silly implementation detail.

**Pull request merged**